File: sdks/python/starfish/presence.py
from __future__ import annotations

import json
from typing import Any

from .connection import Connection
from .emitter import Observable
from .id import next_id
from .limits import MAX_PRESENCE_SIZE, validate_payload_size
from .session import Session
from .types import StarfishFrame, StarfishHeader


class Presence:
    def __init__(self, connection: Connection, session: Session) -> None:
        self._connection = connection
        self._session = session
        self._presence_map: dict[str, Any] = {}

        self.presence: Observable[dict[str, Any]] = Observable({})

    def set(self, payload: Any) -> None:
        session_name = self._require_session()
        validate_payload_size(json.dumps(payload), MAX_PRESENCE_SIZE, "Presence payload")

        frame = StarfishFrame(
            header=StarfishHeader(
                id=next_id("pres"),
                resource="presence",
                method="set",
                kind="request",
                session=session_name,
            ),
            payload=payload,
        )

        import asyncio

        asyncio.ensure_future(self._connection.send(frame))

    async def set_async(self, payload: Any) -> None:
        session_name = self._require_session()
        validate_payload_size(json.dumps(payload), MAX_PRESENCE_SIZE, "Presence payload")

        frame = StarfishFrame(
            header=StarfishHeader(
                id=next_id("pres"),
                resource="presence",
                method="set",
                kind="request",
                session=session_name,
            ),
            payload=payload,
        )

        await self._connection.send(frame)

    def handle_frame(self, frame: StarfishFrame) -> None:
        if (
            frame.header.resource == "presence"
            and frame.header.method == "updated"
            and frame.header.from_
        ):
            self._presence_map[frame.header.from_] = frame.payload
            self.presence.set(dict(self._presence_map))

    def clear(self) -> None:
        self._presence_map.clear()
        self.presence.set({})

    def _require_session(self) -> str:
        session = self._session.current
        if not session:
            raise RuntimeError("Not in a session. Call join() first.")
        return session

File: sdks/golang/starfish/presence.go
package starfish

import (
	"context"
	"encoding/json"
	"sync"
	"time"
)

// presenceManager handles presence set and tracking.
type presenceManager struct {
	mu      sync.RWMutex
	conn    *connection
	idg     *IDGenerator
	session func() string
	state   map[string]map[string]any // clientID → presence data
}

func newPresenceManager(conn *connection, idg *IDGenerator, session func() string) *presenceManager {
	return &presenceManager{
		conn:    conn,
		idg:     idg,
		session: session,
		state:   make(map[string]map[string]any),
	}
}

// set updates the client's presence.
func (p *presenceManager) set(ctx context.Context, payload map[string]any) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	if err := ValidatePayloadSize(data, MaxPresenceSize, "presence"); err != nil {
		return err
	}

	ts := time.Now().UnixMilli()
	frame := &Frame{
		Header: Header{
			V:        1,
			ID:       p.idg.Next("pres"),
			Resource: "presence",
			Method:   "set",
			Kind:     "request",
			Session:  p.session(),
			Ts:       &ts,
		},
		Payload: payload,
	}

	return p.conn.send(ctx, frame)
}

// handleFrame processes presence.updated events.
func (p *presenceManager) handleFrame(f *Frame) {
	if f.Header.Resource != "presence" || f.Header.Method != "updated" {
		return
	}
	from := f.Header.From
	if from == "" {
		return
	}

	p.mu.Lock()
	p.state[from] = f.Payload
	p.mu.Unlock()
}

// get returns the presence data for a given client.
func (p *presenceManager) get(clientID string) map[string]any {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.state[clientID]
}

// getAll returns a copy of the full presence map.
func (p *presenceManager) getAll() map[string]map[string]any {
	p.mu.RLock()
	defer p.mu.RUnlock()
	out := make(map[string]map[string]any, len(p.state))
	for k, v := range p.state {
		out[k] = v
	}
	return out
}

// clear resets the presence state.
func (p *presenceManager) clear() {
	p.mu.Lock()
	p.state = make(map[string]map[string]any)
	p.mu.Unlock()
}
